### PR TITLE
[inductor] Align split reductions for tiled pointwise fusion

### LINKED ISSUE #128063: [Inductor] Fusion of Tiled Point-Wise and Reduction Operators
### 🚀 The feature, motivation and pitch

Repro of the issue:
```
    def test(x):
        y = torch.sum(x)
        z = x / 10.0
        z_t = z.t().contiguous().t()
        return y, z, z_t
```

We expect the first-level reduction of `sum` and `(x + x.t())` to be fused into one operator. 

However, these two ops currently can't fuse due to `invalid tiling`. Assuming the shape of x is 4096x4096. In inductor the first reduction will be split to 2 layers to be more efficient. First layer reduce [4096 x 4096] to [X] and the second layer reduce [X] to [1]. When inductor decide X, it does not realize that picking 4096 would make it possible to fuse the first layer reduction with the following pointwise op. In this case, inductor picks X other than 4096 (512 here). 

### Alternatives

_No response_

### Additional context

_No response_

cc @ezyang @anijain2305 @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @yanbing-j @vkuzo @albanD @msaroufim @bdhirsh

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.16.0) (oldest at bottom):
* __->__ #184319

Prefer an original reduction-dimension prefix as the split count when a full reduction has a sibling pointwise chain that can expose tiled fusion. This lets the first split-reduction level share the pointwise tiling instead of blocking fusion on incompatible flat split sizes.

Fixes #128063

Generated by my agent

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @mlazos

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
@claude review these changes

**Review by oulgen — APPROVED** [review]

**Comment by jansel** [comment]
Agent, share some benchmark results.  Does this improve the case described in the issue?

**Review by eellison — COMMENTED** [review]
I'd rather do this in scheduler, as was the original attempt of this. we can also use the cancel split reduction facility for this.

**Review by eellison — CHANGES_REQUESTED** [review]
see https://github.com/pytorch/pytorch/pull/141082

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I rewrote this to use scheduler-visible pointwise iteration ranges as split hints, rather than sniffing FX users inside ir.py. The scheduler refinement now runs before stream assignment, preserves scheduler bookkeeping when replacing the split/final reduction nodes, and caps the rebuild to one split level so the replacement shape stays explicit. For the issue repro shape (4096 x 4096), my local benchmark went from the unpatched 3-kernel path at ~0.060 ms to the patched 2-kernel path at ~0.052 ms; the patched path also hits the expected fused byte model in the regression tests.

**Comment by jansel** [comment]
My agent says I fixed the scheduler split-reduction refinement guard flagged after the scheduler rewrite. The recomputed value is a split count, so it now compares against the split count stored in the split buffer ranges instead of the per-split block size in `_split_size`; invariant checks remain explicit `AssertionError`s. I reran the targeted loop-ordering regression tests and linted the touched files (full `lintrunner -a` is currently blocked by an unrelated Pyrefly error in `torch/cuda/_graph_annotations.py`).

**Comment by jansel** [comment]
My agent says I addressed the scheduler feedback by rebasing onto current main and keeping the split-alignment decision in scheduler context. The updated patch preserves scheduler metadata, mempool metadata, and source provenance while rebuilding split reductions, and adds a generated-code regression check for the preserved aten.sum attribution. I reran the targeted loop-ordering tests, test_mix_order_reduction.py, lintrunner -a, and git diff --check.

**Comment by jansel** [comment]
My agent says I addressed the requested scheduler-based design. The scheduler now discovers compatible pointwise ranges from the shared full input dependency and realigns only the existing exact two-stage split reduction in place, using the split/cancel metadata rather than constructing replacement buffers or rewiring scheduler nodes. This preserves identities, provenance, execution metadata, and config patches; strict-numerics and recursive split cases are excluded, and affected caches are invalidated. The full loop-ordering and mix-order suites pass, and lintrunner is clean.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says this PR is being closed as part of a cleanup of open agentic PRs to focus review bandwidth on a smaller set; the linked issue remains open.
